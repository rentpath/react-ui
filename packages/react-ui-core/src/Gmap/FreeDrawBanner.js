import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed'
import classnames from 'classnames'

@themed(/^Gmap_FreeDrawBanner/, { pure: true })
export default class FreeDrawBanner extends PureComponent {
  static propTypes = {
    theme: PropTypes.object,
    className: PropTypes.string,
    bannerText: PropTypes.string,
    bannerType: PropTypes.string,
  }

  static defaultProps = {
    theme: {},
  }

  render() {
    const {
      theme,
      className,
      bannerText,
      bannerType,
      ...rest
    } = this.props

    return (
      <div
        className={classnames(
          className,
          theme[`Gmap_FreeDrawBanner-${bannerType}`],
        )}
        {...rest}
      >
        {bannerText}
      </div>
    )
  }
}
