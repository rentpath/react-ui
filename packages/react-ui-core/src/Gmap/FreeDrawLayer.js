import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import themed from 'react-themed'
import { GMAP_EVENTS } from './utils/const'
import { setupEvents, removeEvent } from './utils/mapEventHelpers'

const enableControls = (enable = true) => ({
  draggable: enable,
  zoomControl: enable,
  scrollwheel: enable,
  disableDoubleClickZoom: enable,
})

@themed(['Gmap_FreeDrawLayer'], { pure: true })
export default class FreeDrawLayer extends PureComponent {
  static propTypes = {
    map: PropTypes.object,
    onMapDrawStart: PropTypes.func,
    onMapDrawEnd: PropTypes.func,
    theme: PropTypes.object,
    mapControls: PropTypes.object,
    shapes: PropTypes.object,
    multipleShapes: PropTypes.bool,
    disabled: PropTypes.bool,
    dataStyle: PropTypes.object,
  }

  static defaultProps = {
    theme: {},
    map: {},
    multipleShapes: false,
    disabled: false,
    shapes: {},
    dataStyle: {},
  }

  constructor(props) {
    super(props)
    this.state = { drawing: false }
    this.addShapes()
  }

  componentDidUpdate(prevProps) {
    if (this.props.shapes !== prevProps.shapes) {
      this.clearAllShapes()
      this.addShapes()
    }
  }

  getPolygonCoords(poly) {
    return poly.getPath().getArray().map(path => [path.lng(), path.lat()])
  }

  enableMapControls() {
    const { map, mapControls } = this.props
    map.setOptions({
      ...enableControls(true),
      ...mapControls,
    })
    this.setState({
      drawing: false,
    })
  }

  disableMapControls() {
    const { map } = this.props
    map.setOptions(enableControls(false))
    this.setState({ drawing: true })
  }

  @autobind
  enableMapDraw() {
    const { map, multipleShapes } = this.props
    this.disableMapControls()

    if (!multipleShapes) {
      this.clearAllShapes()
    }
    this.events = setupEvents(map, GMAP_EVENTS, {
      onMouseDown: this.drawFreeHand,
    }, true)
  }

  createPolyline() {
    const { map, dataStyle } = this.props

    return new window.google.maps.Polyline({
      map,
      clickable: false,
      ...dataStyle.polyline,
    })
  }

  createPolygon(polyline, attributes) {
    const { map } = this.props

    return new window.google.maps.Polygon({
      map,
      path: polyline,
      clickable: false,
      ...attributes,
    })
  }

  @autobind
  drawFreeHand() {
    const { map, dataStyle, onMapDrawStart, onMapDrawEnd } = this.props
    const polyline = this.createPolyline()
    const polygon = this.createPolygon(polyline.getPath(), dataStyle.polylineFill)

    if (onMapDrawStart) onMapDrawStart()

    const listener = setupEvents(map, GMAP_EVENTS, {
      onMouseMove: e => polyline.getPath().push(e.latLng),
      onTouchStart: e => polyline.getPath().push(e.latLng),
    }, false)

    const onMouseUp = () => {
      removeEvent(listener.onMouseMove)
      polyline.setMap(null)
      polygon.setMap(null)

      if (onMapDrawEnd) onMapDrawEnd(this.getPolygonCoords(polygon))
      this.enableMapControls()
    }

    this.events = setupEvents(map, GMAP_EVENTS, {
      onMouseUp,
      onTouchEnd: onMouseUp,
    }, true)
  }

  addShapes() {
    const shapes = this.formatLongLatToGmapsCoordinates()
    this.polygons = shapes.map(path =>
      this.createPolygon(
        path,
        {
          ...this.props.dataStyle.polyline,
          ...this.props.dataStyle.polygon,
        }
      )
    )
  }

  @autobind
  clearAllShapes() {
    this.polygons.forEach(poly => poly.setMap(null))
    this.polygons = []
  }

  formatLongLatToGmapsCoordinates() {
    // Should be refactored later to
    // over(pipe(allArr, allArr), point => ({lat: point[1], lng: point[0]})(Object.values(shape)

    return Object.values(this.props.shapes).map(polygon => (
      polygon.map(point => ({ lat: point[1], lng: point[0] }))
    ))
  }

  render() {
    const { theme, disabled } = this.props
    const { drawing } = this.state
    // This button is temporary - will be replaced with the tool tip when built
    return (
      <button
        className={theme.Gmap_FreeDrawLayer}
        onClick={this.enableMapDraw}
        disabled={disabled || drawing}
      >
        Click to draw
      </button>
    )
  }
}
