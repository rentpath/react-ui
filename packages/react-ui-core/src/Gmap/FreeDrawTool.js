import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed'
import classnames from 'classnames'

@themed(/^Gmap_FreeDrawTool/, { pure: true })
export default class FreeDrawTool extends PureComponent {
  static propTypes = {
    theme: PropTypes.object,
    className: PropTypes.string,
    buttonText: PropTypes.string,
    status: PropTypes.oneOf(['active', 'disabled', 'clicked']),
    handleClick: PropTypes.func,
  }

  static defaultProps = {
    theme: {},
    buttonText: 'Draw Search',
    status: 'active',
  }

  render() {
    const {
      theme,
      className,
      buttonText,
      status,
      handleClick,
      ...rest
    } = this.props

    return (
      <button
        className={classnames(
          className,
          theme[`Gmap_FreeDrawTool-${status}`],
        )}
        onClick={handleClick}
        {...rest}
      >
        {buttonText}
      </button>
    )
  }
}
