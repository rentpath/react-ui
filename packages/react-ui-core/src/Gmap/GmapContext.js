import React from 'react'

export const GmapContext = React.createContext('googleMap')
