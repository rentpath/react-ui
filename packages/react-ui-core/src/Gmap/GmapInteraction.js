import once from 'lodash/once'
import get from 'lodash/fp/get'
import isArray from 'lodash/isArray'
import bind from 'lodash/bind'
import includes from 'lodash/fp/includes'
import { setupMarker, removeMarker } from './utils/markerHelpers'

class GmapCallbackFactory {
  constructor() {
    let map = {}
    /* eslint-disable consistent-return */
    this.registerMap = once(newMap => {
      map = newMap
    })

    this.call = (f, args) => {
      // We need to bind non-nested functions, as google uses this.get internally
      if (get(f)(map)) {
        const mapApiFunc = includes('.')(f) ? get(f)(map) : bind(get(f)(map), map)

        if (mapApiFunc && typeof mapApiFunc === 'function') {
          if (isArray(args)) {
            return mapApiFunc(...args)
          }
          return mapApiFunc(args)
        }
      }

      return undefined
    }

    this.MarkerInteraction = {
      setupMarker: props => setupMarker(map, props),
      removeMarker,
    }
  }
}

export const GmapInteraction = new GmapCallbackFactory()
