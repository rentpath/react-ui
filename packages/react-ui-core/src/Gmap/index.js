export { default } from './Gmap'
export { default as Gmap } from './Gmap'
export { default as GmapSpinner } from './GmapSpinner'
export { default as InfoWindow } from './InfoWindow'
export { default as Layer } from './Layer'
export { default as Marker } from './Marker'
export { default as Markers } from './Markers'
export { default as OverlayView } from './OverlayView'
export { GmapInteraction } from './GmapInteraction'
export { setupMarker, removeMarker } from './utils/markerHelpers'
export { default as FreeDrawLayer } from './FreeDrawLayer'
