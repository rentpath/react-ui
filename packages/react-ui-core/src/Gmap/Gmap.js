import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed'
import classnames from 'classnames'
import startsWith from 'lodash/startsWith'
import pickBy from 'lodash/pickBy'
import { GmapInteraction } from './GmapInteraction'
import withGoogleScript from './withGoogleScript'
import { setupEvents } from './utils/mapEventHelpers'

const DEFAULT_ZOOM = 8
const DEFAULT_MIN_ZOOM = 5
const DEFAULT_CENTER = { // Atlanta, GA
  lat: 33.7490,
  lng: -84.3880,
}

const EVENTS = {
  onBoundsChanged: 'bounds_changed',
  onCenterChanged: 'center_changed',
  onClick: 'click',
  onDoubleClick: 'dblclick',
  onDrag: 'drag',
  onDragEnd: 'dragend',
  onDragStart: 'dragstart',
  onHeadingChanged: 'heading_changed',
  onIdle: 'idle',
  onMaptypeIdChanged: 'maptypeid_changed',
  onMouseMove: 'mousemove',
  onMouseOut: 'mouseout',
  onMouseOver: 'mouseover',
  onProjectionChanged: 'projection_changed',
  onResize: 'resize',
  onRightClick: 'rightclick',
  onTilesLoaded: 'tilesloaded',
  onTiltChanged: 'tilt_changed',
  onZoomChanged: 'zoom_changed',
}

const EVENT_NAMES = Object.keys(EVENTS)

const MAP_CONTROLS = {
  fullscreenControl: false,
  streetViewControl: false,
  zoomControl: true,
  mapTypeControl: true,
  scaleControl: true,
  rotateControl: true,

}

@themed(/^Gmap/, { pure: true })

export class Gmap extends PureComponent {
  static propTypes = {
    theme: PropTypes.object,
    className: PropTypes.string,
    children: PropTypes.node,
    center: PropTypes.shape({
      lat: PropTypes.number,
      lng: PropTypes.number,
    }),
    zoom: PropTypes.number,
    minZoom: PropTypes.number,
    maxZoom: PropTypes.number,
    clickableIcons: PropTypes.bool,
    mapControls: PropTypes.shape({
      fullscreenControl: PropTypes.bool,
      mapTypeControl: PropTypes.bool,
      zoomControl: PropTypes.bool,
      streetViewControl: PropTypes.bool,
      scaleControl: PropTypes.bool,
      rotateControl: PropTypes.bool,
      draggable: PropTypes.bool,
      scrollwheel: PropTypes.bool,
      disableDoubleClickZoom: PropTypes.bool,
    }),
    stylingFunction: PropTypes.func,
  }

  static defaultProps = {
    theme: {},
    zoom: DEFAULT_ZOOM,
    minZoom: DEFAULT_MIN_ZOOM,
    center: DEFAULT_CENTER,
    clickableIcons: false,
    stylingFunction: () => ({}),
  }

  constructor(props) {
    super(props)

    this.state = { map: null }
    this.googleMap = React.createRef()
  }

  componentDidMount() {
    this.initMap()
  }

  componentDidUpdate(prevProps) {
    const { center, zoom } = this.props

    if (this.map) {
      if (this.isCenterChange(prevProps.center, center)) {
        this.map.setCenter(center)
      }

      if (zoom && prevProps.zoom !== zoom) {
        this.map.setZoom(zoom)
      }
    }
  }

  componentWillUnmount() {
    if (this.map) {
      window.google.maps.event.clearInstanceListeners(this.map)
    }
  }

  get children() {
    const { map } = this.state
    const children = React.Children.toArray(this.props.children)
    const props = { map }

    return React.Children.map(children, child =>
      React.cloneElement(child, props))
  }

  get mapControls() {
    return { ...MAP_CONTROLS, ...this.props.mapControls }
  }

  initMap() {
    const {
      zoom,
      center,
      minZoom,
      maxZoom,
      clickableIcons,
      stylingFunction,
    } = this.props

    this.map = new window.google.maps.Map(this.googleMap.current, {
      zoom,
      clickableIcons,
      minZoom,
      maxZoom,
      center,
      ...this.mapControls,
    })

    this.map.data.setStyle(stylingFunction)

    setupEvents(this.map, EVENTS, this.props)

    this.setState({ map: this.map })
    GmapInteraction.registerMap(this.map)
  }

  isCenterChange(prev, next) {
    const { lat: plat, lng: plng } = prev || {}
    const { lat: nlat, lng: nlng } = next || {}

    if (plat && plng && !(nlat && nlng)) return false
    return plat !== nlat || plng !== nlng
  }

  render() {
    const { map } = this.state
    const {
      theme,
      className,
      children,
      ...rest
    } = this.props

    const attributes = pickBy(rest, (_, key) => startsWith(key, 'data'))

    return (
      <div
        ref={this.googleMap}
        className={classnames(
          theme.Gmap,
          className,
        )}
        data-tid="google-map"
        {...attributes}
      >
        {map && this.children}
      </div>
    )
  }
}

EVENT_NAMES.forEach(name => {
  Gmap.propTypes[name] = PropTypes.func
})

export default withGoogleScript(Gmap)
