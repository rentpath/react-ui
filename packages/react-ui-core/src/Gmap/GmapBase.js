import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed'
import classnames from 'classnames'
import startsWith from 'lodash/startsWith'
import pickBy from 'lodash/pickBy'
import { GmapContext } from './GmapContext'
import { GmapInteraction } from './GmapInteraction'

const DEFAULT_ZOOM = 8
const DEFAULT_CENTER = { // Atlanta, GA
  lat: 33.7490,
  lng: -84.3880,
}

const EVENTS = {
  onBoundsChanged: 'bounds_changed',
  onCenterChanged: 'center_changed',
  onClick: 'click',
  onDoubleClick: 'dblclick',
  onDrag: 'drag',
  onDragEnd: 'dragend',
  onDragStart: 'dragstart',
  onHeadingChnaged: 'heading_changed',
  onIdle: 'idle',
  onMaptypeIdChanged: 'maptypeid_changed',
  onMouseMove: 'mousemove',
  onMouseOut: 'mouseout',
  onMouseOver: 'mouseover',
  onProjectionChanged: 'projection_changed',
  onResize: 'resize',
  onRightclick: 'rightclick',
  onTilesLoaded: 'tilesloaded',
  onTiltChanged: 'tilt_changed',
  onZoomChanged: 'zoom_changed',
}

const EVENT_NAMES = Object.keys(EVENTS)

@themed(/^Gmap/, { pure: true })

class GmapBase extends PureComponent {
  static propTypes = {
    theme: PropTypes.object,
    className: PropTypes.string,
    children: PropTypes.node,
    center: PropTypes.shape({
      lat: PropTypes.number,
      lng: PropTypes.number,
    }),
    zoom: PropTypes.number,
    clickableIcons: PropTypes.bool,
    mapControls: PropTypes.shape({
      fullscreenControl: PropTypes.bool,
      mapTypeControl: PropTypes.bool,
      zoomControl: PropTypes.bool,
      streetViewControl: PropTypes.bool,
    }),
  }

  static defaultProps = {
    theme: {},
    zoom: DEFAULT_ZOOM,
    clickableIcons: false,
    mapControls: {
      fullscreenControl: false,
      streetViewControl: false,
    },
  }

  constructor(props) {
    super(props)

    this.state = { map: null }
    this.googleMap = React.createRef()
  }

  componentDidMount() {
    this.initMap()
  }

  componentWillUnmount() {
    if (this.map) {
      window.google.maps.event.clearInstanceListeners(this.map)
    }
  }

  setupEvents() {
    EVENT_NAMES.forEach(name => {
      if (this.props[name]) {
        this.map.addListener(EVENTS[name], this.handleEvent(name))
      }
    })
  }

  initMap() {
    const {
      zoom,
      mapControls,
      center,
      clickableIcons,
    } = this.props

    this.map = new window.google.maps.Map(this.googleMap.current, {
      zoom,
      clickableIcons,
      center: center || DEFAULT_CENTER,
      ...mapControls,
    })

    this.setupEvents()
    this.setState({ map: this.map })
    GmapInteraction.registerMap(this.map)
  }

  handleEvent(name) {
    return () => {
      this.props[name](this.map)
    }
  }

  render() {
    const { map } = this.state
    const {
      theme,
      className,
      children,
      ...rest
    } = this.props

    const attributes = pickBy(rest, (_, key) => startsWith(key, 'data'))

    return (
      <div
        ref={this.googleMap}
        className={classnames(
          theme.Gmap,
          className,
        )}
        data-tid="google-map"
        {...attributes}
      >
        <GmapContext.Provider value={{ map }}>
          {map && children}
        </GmapContext.Provider>
      </div>
    )
  }
}

EVENT_NAMES.forEach(name => {
  GmapBase.propTypes[name] = PropTypes.func
})

export default GmapBase
