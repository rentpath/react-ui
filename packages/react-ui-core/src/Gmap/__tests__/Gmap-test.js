import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { Gmap as ThemedGmap } from '../Gmap'

const Gmap = ThemedGmap.WrappedComponent

const theme = {
  Gmap: 'mapClass',
}

describe('Gmap', () => {
  it('applies a custom className', () => {
    const wrapper = shallow(<Gmap className="customMap" />)
    expect(wrapper.prop('className')).toEqual('customMap')
  })

  it('applies a theme className', () => {
    const wrapper = shallow(<Gmap theme={theme} />)
    expect(wrapper.prop('className')).toEqual('mapClass')
  })

  it('applies extra props correctly', () => {
    const wrapper = shallow(<Gmap data-tag_section="map-tag" />)
    expect(wrapper.prop('data-tag_section')).toEqual('map-tag')
  })

  it('sets up only those events that have been passed as props', () => {
    const mapSpy = jest.spyOn(window.google.maps.event, 'addListener')
    shallow(
      <Gmap
        onClick={jest.fn()}
        onBoundsChanged={jest.fn()}
      />
    )

    expect(mapSpy).toHaveBeenCalledWith(expect.any(Object), 'bounds_changed', expect.any(Function))
    expect(mapSpy).toHaveBeenCalledWith(expect.any(Object), 'click', expect.any(Function))
    expect(mapSpy).not.toHaveBeenCalledWith(expect.any(Object), 'mouseout', expect.any(Function))
  })

  it('updates the map zoom when zoom prop changes', () => {
    const wrapper = shallow(<Gmap zoom={10} />)
    const setZoom = jest.fn()
    const instance = wrapper.instance()
    instance.map = {
      ...instance.map,
      setZoom,
    }
    const testZoom = 5
    wrapper.setProps({ zoom: testZoom })
    expect(setZoom).toHaveBeenCalledWith(testZoom)
  })

  it('updates the map center when center prop changes', () => {
    const wrapper = shallow(<Gmap center={{ lat: 33.7490, lng: -84.3880 }} />)
    const setCenter = jest.fn()
    const instance = wrapper.instance()
    instance.map = {
      ...instance.map,
      setCenter,
    }
    const testCenter = {
      lat: 34.0701,
      lng: -118.3497,
    }
    wrapper.setProps({ center: testCenter })
    expect(setCenter).toHaveBeenCalledWith(testCenter)
  })

  describe('clones children correctly', () => {
    it('renders only children that are not null / undefined and passes map as a prop ', () => {
      const snap = renderer
        .create(
          <Gmap>
            <div />
            {null}
            <div />
          </Gmap>
        )
        .toJSON()
      expect(snap).toMatchSnapshot()
    })

    it('renders no children when none defined', () => {
      const snap = renderer.create(<Gmap />).toJSON()
      expect(snap).toMatchSnapshot()
    })
  })
})
