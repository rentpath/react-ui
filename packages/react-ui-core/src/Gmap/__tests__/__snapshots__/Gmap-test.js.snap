// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gmap clones children correctly renders no children when none defined 1`] = `
<div
  className=""
  data-tid="google-map"
/>
`;

exports[`Gmap clones children correctly renders only children that are not null / undefined and passes map as a prop  1`] = `
<div
  className=""
  data-tid="google-map"
>
  <div
    map={
      Object {
        "addListener": [Function],
        "data": Object {
          "addGeoJson": [Function],
          "loadGeoJson": [Function],
          "setStyle": [Function],
        },
      }
    }
  />
  <div
    map={
      Object {
        "addListener": [Function],
        "data": Object {
          "addGeoJson": [Function],
          "loadGeoJson": [Function],
          "setStyle": [Function],
        },
      }
    }
  />
</div>
`;
